presto-main/src/main/java/com/facebook/presto/operator/SharedPartitionedMemoryContext.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+public class SharedPartitionedMemoryContext
+{
+    private int reservations = 0;
+    private Optional<Integer> partitionKey;
+    private Optional<Long> inMemorySizeInBytes;
+    private Optional<TaskContext> taskContext;
+
+    public synchronized void reserve(int partitionKey, OperatorContext operatorContext, long inMemorySizeInBytes)
+    {
+        if (this.reservations == 0) {
+            this.partitionKey = Optional.of(partitionKey);
+            this.inMemorySizeInBytes = Optional.of(inMemorySizeInBytes);
+            TaskContext taskContext = operatorContext.getDriverContext().getPipelineContext().getTaskContext();
+            this.taskContext = Optional.of(taskContext);
+            operatorContext.reserveMemory(inMemorySizeInBytes);
+            operatorContext.transferMemoryToTaskContext(inMemorySizeInBytes);
+        }
+        else {
+            checkCurrentReservationPresent();
+            checkArgument(this.partitionKey.get() == partitionKey && this.inMemorySizeInBytes.get() == inMemorySizeInBytes,
+                    "Inconsistent memory reservation. " + getInconsistenceDescription(partitionKey, inMemorySizeInBytes));
+        }
+        reservations++;
+    }
+
+    public synchronized void free(int partitionKey, long inMemorySizeInBytes)
+    {
+        checkArgument(reservations > 0, format(
+                "Inconsistent memory de-reservation. No current reservation present, " +
+                "requested free for %s bytes for partition %s.",
+                partitionKey,
+                inMemorySizeInBytes
+        ));
+        checkCurrentReservationPresent();
+        checkArgument(this.partitionKey.get() == partitionKey && this.inMemorySizeInBytes.get() == inMemorySizeInBytes,
+                "Inconsistent memory de-reservation. " + getInconsistenceDescription(partitionKey, inMemorySizeInBytes));
+        reservations--;
+        if (reservations == 0) {
```

Comment:
It's not possible ATM due to PartitionedConsumption coordinating the partions consumption :D Also, if this were to happen - we'd just have a little bit less effective sanity check.